File: backend/app/models/data_upload.py
from datetime import datetime
from enum import Enum
from typing import Optional, Annotated

from bson import ObjectId
from pydantic import BaseModel, Field, BeforeValidator


class Status(str, Enum):
    PENDING = "Pending"
    VALIDATING = "Validating"
    FAILED = "Failed"
    COMPLETE = "Complete"


PyObjectId = Annotated[
    str,
    BeforeValidator(lambda v: str(v) if isinstance(v, ObjectId) else v),
]


class DataUpload(BaseModel):
    id: Optional[PyObjectId] = Field(None, alias="_id")
    goal_id: str
    uploader_id: str
    file_name: str
    file_path: str
    upload_timestamp: datetime
    status: Status
    insights: Optional[str] = None

    class Config:
        from_attributes = True
        populate_by_name = True

File: frontend/src/types/data-upload.ts
export interface DataUpload {
  _id: string;
  goal_id: string;
  uploader_id: string;
  file_name: string;
  file_path: string;
  upload_timestamp: string;
  status: "Pending" | "Validating" | "Failed" | "Complete";
  insights?: string;
}